html/includes/authentication/active_directory.inc.php

```diff
@@ -37,33 +37,43 @@ function authenticate($username, $password)
             // group membership in one of the configured groups is required
             if (isset($config['auth_ad_require_groupmembership']) &&
                 $config['auth_ad_require_groupmembership']) {
-                $search = ldap_search(
-                    $ldap_connection,
-                    $config['auth_ad_base_dn'],
-                    get_auth_ad_user_filter($username),
-                    array('memberOf')
-                );
-                $entries = ldap_get_entries($ldap_connection, $search);
-                unset($entries[0]['memberof']['count']); //remove the annoying count
-
-                foreach ($entries[0]['memberof'] as $entry) {
-                    $group_cn = get_cn($entry);
-                    if (isset($config['auth_ad_groups'][$group_cn]['level'])) {
-                        // user is in one of the defined groups
+                foreach ($config['auth_ad_groups'] as $group => $level) {
+                    // get DN for auth_ad_group
+                    $filter = "(&(objectClass=group)(cn=$group))";
+                    $search = ldap_search(
+                        $ldap_connection,
+                        $config['auth_ad_base_dn'],
+                        "(&(objectClass=group)(cn=$group))",
+                        array("cn")
+                    );
+                    $result = ldap_first_entry($ldap_connection, $search);
+                    $group_dn = ldap_get_dn($ldap_connection, $result);
+
+                    $search = ldap_search(
+                        $ldap_connection,
+                        $config['auth_ad_base_dn'],
+                        // add 'LDAP_MATCHING_RULE_IN_CHAIN to the user filter to search for $username in nested $group_dn
+                        // using "DN" just to get a shorter array back
+                        "(&" . get_auth_ad_user_filter($username) . "(memberOf:1.2.840.113556.1.4.1941:=$group_dn))",
+                        array("DN")
+                    );
+                    $entries = ldap_get_entries($ldap_connection, $search);
+                    unset($entries[0]['memberof']['count']); //remove the annoying count
+
+                    if ($entries["count"] > 0) {
+                        // user is in the current group
                         return true;
                     }
                 }
 
                 // failed to find user
                 if (isset($config['auth_ad_debug']) && $config['auth_ad_debug']) {
                     if ($entries['count'] == 0) {
-                        throw new AuthenticationException('No groups found for user, check base dn');
-                    } else {
```

Comment:
I'm not searching what group(s) a user directly belongs to, but look at the defined groups and check if the user is a member of the listed groups or subgroups of that group. As far as I know "count" is always 0 or positive. On that search, if it's positive, the user is a member of that group and we've already returned "true", which only leaves "0" as a possibility.
Maybe the message should be more precise, like "User is not in one of the required groups, or the defined groups are outside the base dn" ?